Replace total-pid-count waits with targeted waits

neilguertincollabora opened the pull request:
Wait for doc kits or spare kit counts specifically. This removes the chance of
race conditions between waiting for doc kits to shutdown and the spare kit to
start back up.

Follow up to https://github.com/CollaboraOnline/online/pull/8463, addressing other tests that were using the same helpers.

### Summary
An incomplete list of things changed:
* Consolidated pid-related functions from countcoolkits.hpp and test.cpp to a new file kit_pid_helpers.cpp/hpp
* Rewrote the retry loop
* Added or clarified lots of log messages
* Simplified pid killing in HTTPCrashTest
* Fixed HTTPCrashTest::testBarren
* HTTPCrashTest::testRecoverAfterKitCrash no longer wastes 20s trying then retrying to connect
* HTTPWSTest::testReloadWhileDisconnecting still failing, it passes if you wait for the process to go away but I don't think that's the intent of the test
* Can log pids with kit_pid_helpers::logKitProcesses. This isn't used directly in the tests but is useful for debugging.

Things I specifically want feedback on:
* Names. Is kit_pid_helpers okay? How about the phrases "Doc Kits" and "Spare Kits". Or anything else, just let me know what you would have named them. 
* Makefile. It seems that because the cpp and hpp files are separate I need to add the cpp to the unit test sources. Is there a better way to do that?
** I tried adding kit_pid_helpers.cpp to unittest_SOURCES but I get linking errors for COOLWSD::getKitPids() etc
** I tried adding COOLWSD.cpp to wsd_sources but it complains that there are multiple main functions
** I tired adding the stubs to wsd/TestStubs.cpp but I can't because they're static

Example logs, not exactly representative of the log viewing experience because I only grabbed the TST lines and because this is only one test that doesn't exercise all the functions.
Before:
```
wsd-1467117-1467227 2024-03-07 16:19:59.636363 -0500 [ coolwsd ] TST  countCoolKitProcesses  [countCoolKitProcesses] (+0ms): Waiting until coolkit processes are exactly 1. Coolkits: 1 | countcoolkits.hpp:66
wsd-1467117-1467227 2024-03-07 16:19:59.636367 -0500 [ coolwsd ] TST  countCoolKitProcesses  [countCoolKitProcesses] (+0ms): Test kit pids are [1467195 ] after waiting for 0ms| countcoolkits.hpp:85
wsd-1467117-1467227 2024-03-07 16:19:59.636368 -0500 [ coolwsd ] TST  countHowManyCoolkits  [testCountHowManyCoolkits] (+0ms): Initial coolkit count is 1| countcoolkits.hpp:105
wsd-1467117-1467227 2024-03-07 16:19:59.636369 -0500 [ coolwsd ] TST  countHowManyCoolkits  [testCountHowManyCoolkits] (+0ms): PASS: InitialCoolKitCount > 0 [true]| countcoolkits.hpp:106
wsd-1467117-1467227 2024-03-07 16:19:59.636477 -0500 [ coolwsd ] TST  crashKit  [getDocumentPathAndURL] (+0ms): Test file: /tmp/crashKit_46ccbe28_empty.odt| helpers.hpp:194
wsd-1467117-1467227 2024-03-07 16:19:59.636837 -0500 [ coolwsd ] TST  crashKit  [sendTextFrame] (+0ms): Sending 66 bytes: load url=cool/file%3A%2F%2F%2Ftmp%2FcrashKit_46ccbe28_empty.odt/ws| helpers.hpp:213
wsd-1467117-1467227 2024-03-07 16:20:00.118810 -0500 [ coolwsd ] TST  crashKit  [loadDocAndGetSession] (+482ms): PASS: isLoaded [true]| helpers.hpp:631
wsd-1467117-1467227 2024-03-07 16:20:00.118813 -0500 [ coolwsd ] TST  crashKit  [loadDocAndGetSession] (+482ms): Loaded document [cool/file%3A%2F%2F%2Ftmp%2FcrashKit_46ccbe28_empty.odt/ws].| helpers.hpp:633
wsd-1467117-1467227 2024-03-07 16:20:00.118817 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+482ms): Allowing time for kits to spawn and connect to wsd to get cleanly killed| httpcrashtest.cpp:150
wsd-1467117-1467227 2024-03-07 16:20:01.119029 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+1482ms): Killing coolkit instances.| httpcrashtest.cpp:153
wsd-1467117-1467227 2024-03-07 16:20:01.119099 -0500 [ coolwsd ] TST  killLoKitProcesses  [killPids] (+1482ms): kill pids 2| httpcrashtest.cpp:259
wsd-1467117-1467227 2024-03-07 16:20:01.119106 -0500 [ coolwsd ] TST  killLoKitProcesses  [killPids] (+1482ms): Killing 1467195| httpcrashtest.cpp:266
wsd-1467117-1467227 2024-03-07 16:20:01.119128 -0500 [ coolwsd ] TST  killLoKitProcesses  [killPids] (+1482ms): Killing 1467231| httpcrashtest.cpp:266
wsd-1467117-1467227 2024-03-07 16:20:01.119139 -0500 [ coolwsd ] TST  countCoolKitProcesses  [countCoolKitProcesses] (+1482ms): Waiting until coolkit processes are exactly 0. Coolkits: 2 wsd-1467117-1467230 2024-03-07 16:20:00.731638 -0500 [ docbroker_001 ] TRC  #33: setupPollFds getPollEvents: 0x1| net/Socket.hpp:859
0 | countcoolkits.hpp:66
wsd-1467117-1467227 2024-03-07 16:20:01.894950 -0500 [ coolwsd ] TST  countCoolKitProcesses  [countCoolKitProcesses] (+2258ms): Test kit pids are [] after waiting for 775ms| countcoolkits.hpp:85
wsd-1467117-1467227 2024-03-07 16:20:01.894959 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+2258ms): Reading the error code from the socket.| httpcrashtest.cpp:158
wsd-1467117-1467227 2024-03-07 16:20:01.894967 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+2258ms): Shutting down socket.| httpcrashtest.cpp:165
wsd-1467117-1467227 2024-03-07 16:20:01.894986 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+2258ms): PASS: socket->waitForDisconnection(std::chrono::seconds(5)) [true]| httpcrashtest.cpp:168
wsd-1467117-1467227 2024-03-07 16:20:01.895292 -0500 [ coolwsd ] TST  countCoolKitProcesses  [countCoolKitProcesses] (+0ms): Waiting until coolkit processes are exactly 1. Coolkits: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 wsd-1467117-1467117 2024-03-07 16:20:01.889354 -0500 [ coolwsd ] TRC  ppoll start, timeoutMicroS: 250000 size 0| net/Socket.cpp:348
1 | countcoolkits.hpp:66
wsd-1467117-1467227 2024-03-07 16:20:02.180636 -0500 [ coolwsd ] TST  countCoolKitProcesses  [countCoolKitProcesses] (+285ms): Test kit pids are [1467283 ] after waiting for 285ms| countcoolkits.hpp:85
wsd-1467117-1467227 2024-03-07 16:20:02.180640 -0500 [ coolwsd ] TST  noExtraCoolKitsLeft  [testNoExtraCoolKitsLeft] (+285ms): PASS: InitialCoolKitCount == countNow == [1]| countcoolkits.hpp:115
wsd-1467117-1467227 2024-03-07 16:20:02.180643 -0500 [ coolwsd ] TST  noExtraCoolKitsLeft  [testNoExtraCoolKitsLeft] (+285ms):  (2544ms)| countcoolkits.hpp:120
```

After:
```
wsd-1458824-1458922 2024-03-07 16:18:38.908134 -0500 [ coolwsd ] TST  setUp [waitForKitProcessCount] (+0ms): Waiting for kit process count: Doc Kits: 0  Spare Kits: 1 | kit_pid_helpers.cpp:73
wsd-1458824-1458922 2024-03-07 16:18:38.908137 -0500 [ coolwsd ] TST  setUp [waitForKitProcessCount] (+0ms): Current kit processes: Doc Kits: [] Spare Kits: [1458964, ]| kit_pid_helpers.cpp:83
wsd-1458824-1458922 2024-03-07 16:18:38.908139 -0500 [ coolwsd ] TST  setUp [waitForKitProcessCount] (+0ms): Finished waiting for kit process count: Doc Kits: 0  Spare Kits: 1 | kit_pid_helpers.cpp:104
wsd-1458824-1458922 2024-03-07 16:18:38.908167 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+0ms): Loading document| httpcrashtest.cpp:144
wsd-1458824-1458922 2024-03-07 16:18:38.908248 -0500 [ coolwsd ] TST  crashKit  [getDocumentPathAndURL] (+0ms): Test file: /tmp/crashKit_5654a31b_empty.odt| helpers.hpp:194
wsd-1458824-1458922 2024-03-07 16:18:38.908580 -0500 [ coolwsd ] TST  crashKit  [sendTextFrame] (+0ms): Sending 66 bytes: load url=cool/file%3A%2F%2F%2Ftmp%2FcrashKit_5654a31b_empty.odt/ws| helpers.hpp:213
wsd-1458824-1458922 2024-03-07 16:18:39.059931 -0500 [ coolwsd ] TST  crashKit  [loadDocAndGetSession] (+151ms): PASS: isLoaded [true]| helpers.hpp:631
wsd-1458824-1458922 2024-03-07 16:18:39.059946 -0500 [ coolwsd ] TST  crashKit  [loadDocAndGetSession] (+151ms): Loaded document [cool/file%3A%2F%2F%2Ftmp%2FcrashKit_5654a31b_empty.odt/ws].| helpers.hpp:633
wsd-1458824-1458922 2024-03-07 16:18:39.059950 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+151ms): Allowing time for kit to connect to wsd to get cleanly killed| httpcrashtest.cpp:148
wsd-1458824-1458922 2024-03-07 16:18:40.060128 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+1152ms): Killing coolkit instances.| httpcrashtest.cpp:151
wsd-1458824-1458922 2024-03-07 16:18:40.060204 -0500 [ coolwsd ] TST  crashKit  [killPid] (+1152ms): Killing 1458964| httpcrashtest.cpp:247
wsd-1458824-1458922 2024-03-07 16:18:40.060227 -0500 [ coolwsd ] TST  crashKit  [killPid] (+1152ms): Killing 1459018| httpcrashtest.cpp:247
wsd-1458824-1458922 2024-03-07 16:18:40.060245 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1152ms): Waiting for kit process count: Doc Kits: 0  Spare Kits: 0 | kit_pid_helpers.cpp:73
wsd-1458824-1458922 2024-03-07 16:18:40.060255 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1152ms): Current kit processes: Doc Kits: [1458964, ] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:83
wsd-1458824-1458922 2024-03-07 16:18:40.070358 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1162ms): Current kit processes: Doc Kits: [1458964, ] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:40.080553 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1172ms): Current kit processes: Doc Kits: [1458964, ] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
...
wsd-1458824-1458922 2024-03-07 16:18:40.245112 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1336ms): Current kit processes: Doc Kits: [1458964, ] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:40.255411 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1347ms): Current kit processes: Doc Kits: [1458964, ] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:40.265700 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1357ms): Current kit processes: Doc Kits: [1458964, ] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:40.275933 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1367ms): Current kit processes: Doc Kits: [] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:40.286188 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1378ms): Current kit processes: Doc Kits: [] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:40.296482 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+1388ms): Current kit processes: Doc Kits: [] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
...
wsd-1458824-1458922 2024-03-07 16:18:41.116437 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+2208ms): Current kit processes: Doc Kits: [] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:41.126725 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+2218ms): Current kit processes: Doc Kits: [] Spare Kits: [1459018, ]| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:41.137076 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+2228ms): Current kit processes: Doc Kits: [] Spare Kits: []| kit_pid_helpers.cpp:97
wsd-1458824-1458922 2024-03-07 16:18:41.137120 -0500 [ coolwsd ] TST  crashKit  [waitForKitProcessCount] (+2228ms): Finished waiting for kit process count: Doc Kits: 0  Spare Kits: 0 | kit_pid_helpers.cpp:104
wsd-1458824-1458922 2024-03-07 16:18:41.137131 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+2228ms): Shutting down socket.| httpcrashtest.cpp:162
wsd-1458824-1458922 2024-03-07 16:18:41.137139 -0500 [ coolwsd ] TST  crashKit  [testCrashKit] (+2228ms): PASS: socket->waitForDisconnection(std::chrono::seconds(5)) [true]| httpcrashtest.cpp:165
wsd-1458824-1458922 2024-03-07 16:18:41.137557 -0500 [ coolwsd ] TST  tearDown [waitForKitProcessCount] (+0ms): Waiting for kit process count: Doc Kits: 0 | kit_pid_helpers.cpp:73
wsd-1458824-1458922 2024-03-07 16:18:41.137615 -0500 [ coolwsd ] TST  tearDown [waitForKitProcessCount] (+0ms): Current kit processes: Doc Kits: [] Spare Kits: []| kit_pid_helpers.cpp:83
wsd-1458824-1458922 2024-03-07 16:18:41.137625 -0500 [ coolwsd ] TST  tearDown [waitForKitProcessCount] (+0ms): Finished waiting for kit process count: Doc Kits: 0 | kit_pid_helpers.cpp:104
```


### Checklist

- [X] Code is properly formatted
- [X] All commits have Change-Id
- [X] I have run tests with `make check`
- [X] I have issued `make run` and manually verified that everything looks okay
- [X] Documentation (manuals or wiki) has been updated or is not required

neilguertincollabora commented:
CI is failing because of a git error, I repushed but it is still failing. I did check that all the tests pass with make check, so it is ready for review.

mmeeks commented on test/TileCacheTests.cpp:
```diff
@@ -450,13 +450,14 @@ void TileCacheTests::testDisconnectMultiView()
+        TST_LOG("disconnectMultiView try #" << j);
+
+        // Make sure previous sessions have closed
+        kit_pid_helpers::waitForSpareKit(testname);
```
Superficially this seems not to do what the previous test did surely? surely the previous code waited until there were no more document processes left open (?)

neilguertincollabora replied:
It does the same thing.

countCoolKitProcesses is poorly named because it also waits, in this case for exactly one process.

waitForSpareKit waits until the spare kit is ready but also that there are no doc kits. So exactly one process, but this time it's definitely a spare. Is waitForSpareKit poorly named too? Perhaps. It's mostly used during setup, where you do want to make sure there are no old doc kits hanging around but usually there aren't and you end up waiting for the spare. I had considered a bunch of different names but didn't like any of them and settled on this. I'll think on it some more and see if I can come up with something clearer.

mmeeks replied:
> It does the same thing.

It is now really quite a large patch. You are 100% certain of that, presumably or you wouldn't state this so assertively and affirmatively I assume.

Somewhere along the line here - I think we've dropped a conceptual ball. The 'spare' kit stuff dates back from when we couldn't tell if a process was 'spare' or not - because we saw only the contents of /proc/<pid> So lots of the logic exists to remove the tiresome spare processes from our accounting.

Roll on to today - there is no need for this at all. We should not be waiting for situations where there are 0 spare kits eg. that looks like a nonsense; we should just not be looking at 'spare'-ness at all.

I expected a patch something like this:

diff --git a/test/countcoolkits.hpp b/test/countcoolkits.hpp
index b9c0435308..182cebfc2e 100644
--- a/test/countcoolkits.hpp
+++ b/test/countcoolkits.hpp
@@ -93,7 +93,7 @@ static int countCoolKitProcesses(const int expected,
 // (in the non-ideal event that) it's not 1,
 // it will cause testNoExtraCoolKitsLeft to
 // wait unnecessarily and fail.
-static int InitialCoolKitCount = 1;
+static int InitialCoolKitCount = 0;
 static std::chrono::steady_clock::time_point TestStartTime;
 
 static void testCountHowManyCoolkits()
diff --git a/test/test.cpp b/test/test.cpp
index 59fa4fafbc..0e24685442 100644
--- a/test/test.cpp
+++ b/test/test.cpp
@@ -321,7 +321,7 @@ std::set<pid_t> getForKitPids()
 /// How many live coolkit processes do we have ?
 int getCoolKitProcessCount()
 {
-    return getKitPids().size();
+    return getDocKitPids().size();
 }
 #endif
 

Can you test that instead and see where it falls down; perhaps it does not - perhaps it does.

mmeeks replied:
Quite possibly there is a test or two that really needs to know about spare kits; but the majority don't I expect will need touching.

mmeeks commented on test/kit_pid_helpers.cpp:
```diff
@@ -8,40 +8,24 @@
+#include <lokassert.hpp>
+#include <testlog.hpp>
 
-std::set<pid_t> kit_pid_helpers::getKitPids()
```
If we needed a namespace, and we don't here. It would be cleaner to define the namespace foo { and define functions inside it I think }

neilguertincollabora replied:
The other functions were in the `helpers` namespace, but have `using namespace helpers;` in the test files. I was trying to avoid putting functions in the helpers namespace that aren't defined in helpers.hpp, but I'm not sure if that's actually an issue.

So should I:
- Move these to the helpers namespace
- Keep them in their own (probably renamed to PidHelpers)
- No namespace at all (I don't like this, given there are other functions with the same name in different namespaces, like COOLWSD::getKitPids)

For the implementation in the cpp file, I suppose you could implement them inside brackets as `namespace foo { void bar(){...} }`, but I have never seen it done that way, I have always seen it as `void foo::bar(){...}`. Please confirm this is what you want, I would do it but it feels weird. Maybe I'm confused and it is only necessary for methods?

And then in tests, if using a new namespace, should I:
- `using namespace PidHelpers;`, and call getKitPids()
- or call PidHelpers::getKitPids()

These are style questions, so your call.

mmeeks replied:
on namespaces: cf. $ git grep 'namespace' | grep cpp # :-)
No problem to keep helpers namespace for now, or use PidHelpers if it makes you happy.
Re: invocation - we like the code to be short & readable, without overly-cutsey naming as a general rule.

mmeeks commented on test/TileCacheTests.cpp:
```diff
@@ -19,6 +19,9 @@
 
 #include <cppunit/extensions/HelperMacros.h>
 
+#include <sstream>
```
Formatting changes to your existing code should be squashed into the existing commit. But then I guess for back-porting it's good to have cosmetic cleanup separated I suppose.

neilguertincollabora replied:
It's formatting changes for new code so you're right it's probably better to squash it.

mmeeks reviewed (changes requested):
TBH I'm surprised by the scope of the change. I expected a few lines of change to:

void testCountHowManyCoolkits();
void testNoExtraCoolKitsLeft();

To make them much more robust, and perhaps a few related changes; 10 lines of patch - easy to back-port, move on =)

> Names. Is kit_pid_helpers okay?

That's easy to answer not great, we aim to converge on consistency slowly. So this is the first source code filename with _ in it, it's rather verbose, and its the first namespace with _ in it - so not a great move to rename like that.
Also drop the namespace - this is test code. The rename we're moving towards StldCaps filenames it seems looking at the trend there - so PidHelpers.hpp would be fine if youw ant a rename.

>  How about the phrases "Doc Kits" and "Spare Kits". Or anything else, just let me know what you would have named them.
That's fine :-)

As for the linking problems - I'm a tad confused why linking changed in this regard. Perhaps we can re-try with the very simple solution of having InitialCoolKitCount be zero and improving those functions in-place ?

neilguertincollabora commented:
Updated responding to feedback.
* Squashed commits, moved include changes to first commit
* Renamed kit_pid_helpers.cpp to KitPidHelpers.cpp
* Renamed wait functions to `waitForKitPidsReady` and `waitForKitPidsKilled`
* Moved functions from `kit_pid_helpers::` namespace to `helpers::`
* Modified kill functions in httpcrashtest.cpp

neilguertincollabora commented:
> Hey ho - if you want to go with a monster change instead a minimal improvement we can - why not; but for abundance of clarity -you- are the one debugging any similar problems in this area in future - and also the one back-porting it safely to 23.05. If that's ok - and it passes CI ideally on 23.05 - then happy to approve of course.

I accept that responsibility

mmeeks commented on test/httpcrashtest.cpp:
```diff
@@ -224,57 +214,54 @@ void HTTPCrashTest::testCrashForkit()
-        killLoKitProcesses();
-        countCoolKitProcesses(0);
-        TST_LOG("Communicating after kill.");
-        socket = loadDocAndGetSession(_socketPoll, "empty.odt", _uri, testname);
```
Where did this go ?

neilguertincollabora replied:
As far as I can tell it doesn't do anything. The test is about killing forkit, it does that above and verifies we can still talk to the kit process.

But then it also kills the kit process, which shouldn't be necessary for the test.

Edit: Now I see it's probably verifying that forkit restarts and can spawn new kit processes. I'll re-add it.

mmeeks commented on test/httpcrashtest.cpp:
```diff
@@ -179,28 +177,20 @@ void HTTPCrashTest::testRecoverAfterKitCrash()
-
         std::shared_ptr<http::WebSocketSession> socket2 = loadDocAndGetSession(
             _socketPoll, "empty.odt", _uri, testname, /*isView=*/true, /*isAssert=*/false);
-        if (!socket2)
```
seems reasonable; coolwsd should be up and running reliably in this mode.

neilguertincollabora replied:
Yup, it was wasting 20s by trying to connect 1s before it was ready, then waiting to time out and retry.

mmeeks commented on test/httpwstest.cpp:
```diff
@@ -149,11 +148,6 @@ void HTTPWSTest::testSaveOnDisconnect()
         getResponseMessage(socket1, "pasteresult: success", testname);
 
-        std::set<pid_t> origPids = getDocKitPids();
-        LOK_ASSERT_EQUAL(static_cast<size_t>(1), origPids.size());
```
Why did we loose this ?

neilguertincollabora replied:
Verifying specific pids shouldn't be necessary. Perhaps in one specific test to verify that they don't get reused (is that even possible?) but no need to have lots of extra code to do that in unrelated tests.

It was here because it was the initial fix for the recent unit-httpws failure #8463. If that fix and this one were done together (and they will be backported together) there would have been no need to add it.

mmeeks replied:
The assert checks not so much the pid, but that there is still just one process open; that seems a useful check surely.

neilguertincollabora replied:
The size check was only there to make `*(origPids.begin())` safe. Otherwise it would have gotten a random or null(?) pid.

Like checking that it doesn't reuse pids, checking that it doesn't open two kits could be useful, but that's not what this test is for. I made a note and I'll add those new testpoints later.

mmeeks commented on test/httpwstest.cpp:
```diff
@@ -164,29 +158,15 @@ void HTTPWSTest::testSaveOnDisconnect()
-    }
-    catch (const Poco::Exception& exc)
-    {
-        LOK_ASSERT_FAIL(exc.displayText());
```
Why don't we fail on exception anymore ? that seems important.

neilguertincollabora replied:
It's hard to tell in the diff, but this was two sequential try blocks that I combined into one.

mmeeks commented on test/httpwstest.cpp:
```diff
@@ -164,29 +158,15 @@ void HTTPWSTest::testSaveOnDisconnect()
-        LOK_ASSERT_EQUAL(static_cast<size_t>(1), newPids.size());
-        pid_t newPid = *(newPids.begin());
-        TST_LOG("New pid: " << newPid);
-        LOK_ASSERT_MESSAGE("New pid ("<<newPid<<") should not match old pid ("<<origPid<<")", newPid != origPid);
```
Where did this go ?

neilguertincollabora replied:
See the other comment

mmeeks replied:
Surely this checks that we don't create a new process for the same document when we're re-opening it (?)

mmeeks commented on test/httpwstest.cpp:
```diff
@@ -233,9 +211,6 @@ void HTTPWSTest::testReloadWhileDisconnecting()
         socket = loadDocAndGetSession(_socketPoll, _uri, documentURL, testname);
 
-        // Should have no new instances.
-        LOK_ASSERT_EQUAL(kitcount, countCoolKitProcesses(kitcount));
```
Surely there is a useful check here we lost ?

neilguertincollabora replied:
This test fails and is not running. It passes if you wait between disconnect and reconnect, but since it's called testReloadWhileDisconnecting I think that specifically should not be done. I'll add a comment here explaining that much. When it gets fixed and reenabled we can add whatever waits or asserts we need.

mmeeks replied:
Fine - it's not included in the set currently.

mmeeks reviewed (changes requested):
So looking generally better; a few last bits I don't understand why they are removed.

neilguertincollabora commented:
Just pushed the latest fixes. Here is the diff:
```
diff --git a/test/httpcrashtest.cpp b/test/httpcrashtest.cpp
index 2a6102acc..669581ad3 100644
--- a/test/httpcrashtest.cpp
+++ b/test/httpcrashtest.cpp
@@ -232,6 +232,22 @@ void HTTPCrashTest::testCrashForkit()
         socket->asyncShutdown();
         LOK_ASSERT_MESSAGE("Expected successful disconnection of the WebSocket",
                            socket->waitForDisconnection(std::chrono::seconds(5)));
+
+        TST_LOG("Killing coolkit.");
+        killAllKitProcesses(testname);
+        waitForKitPidsKilled(testname);
+
+        // Forkit should restart
+        waitForKitPidsReady(testname);
+
+        TST_LOG("Communicating after kill.");
+        socket = loadDocAndGetSession(_socketPoll, "empty.odt", _uri, testname);
+        sendTextFrame(socket, "status", testname);
+        assertResponseString(socket, "status:", testname);
+
+        socket->asyncShutdown();
+        LOK_ASSERT_MESSAGE("Expected successful disconnection of the WebSocket",
+                socket->waitForDisconnection(std::chrono::seconds(5)));
     }
     catch (const Poco::Exception& exc)
     {
diff --git a/test/httpwstest.cpp b/test/httpwstest.cpp
index 5c471e36b..06859b65e 100644
--- a/test/httpwstest.cpp
+++ b/test/httpwstest.cpp
@@ -207,6 +207,9 @@ void HTTPWSTest::testReloadWhileDisconnecting()
         LOK_ASSERT_MESSAGE("Expected successful disconnection of the WebSocket",
                            socket->waitForDisconnection(std::chrono::seconds(5)));
 
+        // Do not wait here. Reconnect before disconnect finishes
+        // TODO: Test fails because it is unable to reconnect
+
         // Load the same document and check that the last changes (pasted text) is saved.
         TST_LOG("Loading again.");
         socket = loadDocAndGetSession(_socketPoll, _uri, documentURL, testname);
```

mmeeks reviewed (approved):
Lets get it in then =)
